pkg/proxy/ipvs/proxier.go

```diff
@@ -607,34 +606,29 @@ func NewLinuxKernelHandler() *LinuxKernelHandler {
 // GetModules returns all installed kernel modules.
 func (handle *LinuxKernelHandler) GetModules() ([]string, error) {
 	// Check whether IPVS required kernel modules are built-in
+	var nfconntrack bool
 	kernelVersionStr, err := handle.GetKernelVersion()
 	if err != nil {
 		return nil, err
 	}
-	kernelVersion, err := version.ParseGeneric(kernelVersionStr)
-	if err != nil {
-		return nil, fmt.Errorf("error parsing kernel version %q: %v", kernelVersionStr, err)
-	}
-	ipvsModules := utilipvs.GetRequiredIPVSModules(kernelVersion)
 
-	var bmods, lmods []string
+	ipvsModules := utilipvs.GetRequiredIPVSModules()
+
+	var bmods []string
 
 	// Find out loaded kernel modules. If this is a full static kernel it will try to verify if the module is compiled using /boot/config-KERNELVERSION
 	modulesFile, err := os.Open("/proc/modules")
 	if err == os.ErrNotExist {
-		klog.Warningf("Failed to read file /proc/modules with error %v. Assuming this is a kernel without loadable modules support enabled", err)
-		kernelConfigFile := fmt.Sprintf("/boot/config-%s", kernelVersionStr)
-		kConfig, err := ioutil.ReadFile(kernelConfigFile)
+		bmods, nfconntrack, err := utilipvs.BuiltinIPVSModules(true, kernelVersionStr)
 		if err != nil {
-			return nil, fmt.Errorf("Failed to read Kernel Config file %s with error %v", kernelConfigFile, err)
+			return nil, err
 		}
-		for _, module := range ipvsModules {
-			if match, _ := regexp.Match("CONFIG_"+strings.ToUpper(module)+"=y", kConfig); match {
-				bmods = append(bmods, module)
-			}
+		if !nfconntrack {
```

Comment:
hum, thinking here.

So, the main approach here was actually because in the part of this code we check first if there's a built in statically compiled modules into the kernel.
 
In the BuiltinIPVSModule we have the following:

```
if err != nil && statickernel {
		return nil, false, fmt.Errorf("Failed to read Kernel Config file %s with error %v", modulesConfigFile, err)
```
so it will return false even if there's an error. My thinking here is that returning a warning of nf_conntrack, even if this is false (because an error ocurred first) could lead the user to wrong understanding that kube-proxy is failing because of the lack of nf_conntrack, when actually it's failing because kube-proxy can't read the /boot/config-kernelversion.

So IMO, we could leave the early return of errors before the warning, as the warning is just going to appear if no error happens before while trying to read the modules file :)

WDYT?

Reply:
@rikatz sure, that's make sense.

Reply:
I am re-reading here with a fresh morning the code. I am wondering @rikatz if we need at all nfconntrack variable. 

Example:
```
if !nfconntrack {
                        klog.Warningf("Failed to load nf_conntrack_ipv4 (or nf_conntrack)")
                }
```

This could be popup in BuiltinIPVSModules function. 

Another suggestion, use a switch for this one:
```
if module == KernelModuleNfConntrackIPV4 || module == KernelModuleNfConntrack {
                        nfconntrack = true
                }
```

Reasons for that: 
     You know exactly which module detected 
     Print the message to users with exactly error in case none of the modules detected or just one.
     
"Failed to load nf_conntrack_ipv4 (or nf_conntrack)" seems pretty generic to end users.

Reply:
ok, so my guess here is that this can be a later optimization.

On my first pass, it did made sense to move the boolean to inside BuiltinIPVSModules, but there's a part of the code that we are missing here: the BuiltinIPVSModules is a different function, to get which modules are already loaded as a Builtin (and if nfconntrack modules are there) and then this is validated again further. 

the bool have a default value of false, and can be changed to true by BuiltinIPVSModules OR the later command, that verifies / tries to do a modprobe in those modules.

If we move the message to inside the library, anytime the BuiltinIPVSModules is called and no nfconntrack module is found the message will be issued, even if the modules are later loaded by the modprobe line. So IMO this should stay as is right now, but we can better discuss this!

Reply:
About the switch, as this is almost a one liner if, I'm guessing if using a switch here for a one case only wouldn't be somehow overengineering :D But I might be missing something.

I'm imagining the idea to be something like:

```golang
switch module {
    case utilipvs.LernelModuleNfConntrack, utilipvs.KernelModuleNfConntrackIPV4:
         nfconntrack = true
}
```

Right? So the default case is already covered here (the bool is started as false) so I don't know if it's worth to change this to a switch case :D